### DISCUSSION TITLE
`limactl ls --json` emits a stream of objects instead of an array

### DISCUSSION THREAD

**Discussion by jandubois** [lima-vm/lima#392] [Q&A] [answered]
This makes is harder to parse it using `JSON.parse` from JavaScript:

```console
$ limactl ls --json
{"name":"alpine","status":"Running","dir":"/Users/jan/.lima/alpine","arch":"x86_64","sshLocalPort":60020,"hostAgentPID":16450,"qemuPID":16451}
{"name":"default","status":"Running","dir":"/Users/jan/.lima/default","arch":"x86_64","sshLocalPort":60022,"hostAgentPID":13199,"qemuPID":13200}
```

Is there a reason not to wrap the objects in an array, making the output a valid JSON document?

**Comment by AkihiroSuda** [comment, marked as answer]
I prefer to use [NDJSON](http://ndjson.org/) (aka [JSONlines](https://jsonlines.org/), [LDJSON](https://en.wikipedia.org/wiki/JSON_streaming#Line-delimited_JSON)) everywhere because

- Memory efficient for marshaling large arrays (`limactl ls` does not need to scale for more than 10 elements, though)
- Even if the process crashed, we can get some partial results
- Friendly to shell utilities such as grep. No need to install `gron`/`ungron`.

NDJSON is also used in the event API of the guestagent and the hostagent.

**Comment by AkihiroSuda** [comment]
There seems some libraries for parsing NDJSON/JSONlines/LDJSON with nodejs
- https://www.npmjs.com/package/ndjson
- https://www.npmjs.com/package/json-lines
- https://www.npmjs.com/package/ldjson

**Reply by ghost** [reply]
`jq` also parses NDJSON nicely:
```
limactl list --json | jq
```
